dspy/signatures/utils.py

```diff
@@ -0,0 +1,10 @@
+from typing import Literal
+
+from pydantic.fields import FieldInfo
+
+
+def get_dspy_field_type(field: FieldInfo) -> Literal["input", "output"]:
```

Comment:
I noticed that we use `field.json_schema_extra.get("__dspy_field_type")` in many locations throughout DSPy to fetch the DSPy field type (input or output) from a pydantic field. To simplify this lookup going forward, I figured it would be useful to have a utility.

Happy to file a follow-up PR that replaces existing instances of `field.json_schema_extra.get("__dspy_field_type")` with a call to this utility.

Reply:
Is an error always raised if there's no `dsyp_field_type`?

Reply:
Yeah, the current implementation attempts a dictionary access of the form `field.json_schema_extra["__dspy_field_type"]`, which fails with KeyError if the field is not present

Reply:
ah sorry! - Think this check occurs when you add fields to the signature, `_SignatureMeta._validate_fields` so don't think it's necessary here? Can safely access by key after the signature is created.

Maybe slicker to see if a property can be added to `pydantic.Field` or if `pydantic.Field` can be subclassed.

Reply:
> Maybe slicker to see if a property can be added to pydantic.Field or if pydantic.Field can be subclassed.

I agree - does it make sense to pursue that as a followup to this PR, since it's a broader refactor?

> Think this check occurs when you add fields to the signature, _SignatureMeta._validate_fields so don't think it's necessary here

If this assumption ever fails for some case & we remove this logic, we'll raise a KeyError instead of a pretty error message. Happy to remove it, but not sure it hurts...

Reply:
Yeah agree.